File: bindings/cpp/benches/bench_stream.cpp
/* encrypt_stream_pump throughput vs plaintext size (Streaming
 * Non-AEAD profile) at 1 MiB / 16 MiB / 64 MiB. */

#include <exception>

#include "bench_util.hpp"

int main()
{
    try {
        /* Bench-scale allocation churn leaks Go scratch heap
         * unboundedly without a soft memory cap + aggressive GC; the
         * return values report the previous settings, not an error. */
        (void)itb::set_memory_limit(512LL << 20); /* 512 MiB soft cap */
        (void)itb::set_gc_percent(20);            /* aggressive GC */

        itb::Pipeline pipe = itb::Pipeline::init(
            bench_profile_name("streaming-noaead-triple-v1"),
            bench_build_opts());
        bench_header();
        static const std::size_t sizes[] = {
            std::size_t{1} << 20, std::size_t{16} << 20, std::size_t{64} << 20,
        };
        for (std::size_t size : sizes) {
            std::vector<std::uint8_t> plain(size);
            bench_csprng_fill(plain);
            bench_case("stream_pump", size, [&] {
                (void)pipe.encrypt_stream_pump(itb::as_bytes(plain));
            });
        }
        return 0;
    } catch (const std::exception &e) {
        std::fprintf(stderr, "bench_stream: %s\n", e.what());
        return 1;
    }
}

File: bindings/zig/benches/bench_stream.zig
//! encryptStreamPump throughput vs plaintext size (streaming
//! Non-AEAD profile) at 1 MiB / 16 MiB / 64 MiB.

const std = @import("std");
const itb = @import("itb");
const util = @import("bench_util.zig");

const Ctx = struct {
    pipe: *const itb.Pipeline,
    plain: []const u8,
};

fn runStreamPump(ctx: *const Ctx) !void {
    const wire = try ctx.pipe.encryptStreamPump(ctx.plain);
    ctx.pipe.allocator.free(wire);
}

pub fn main(init: std.process.Init) !void {
    // Bench-scale allocation churn leaks Go scratch heap unboundedly
    // without a soft memory cap + aggressive GC; the return values
    // report the previous settings, not an error.
    _ = itb.setMemoryLimit(512 << 20); // 512 MiB soft cap
    _ = itb.setGcPercent(20); // aggressive GC

    const gpa = init.gpa;
    const arena = init.arena.allocator();
    const env = init.environ_map;

    const opts = try util.buildOpts(arena, env);
    defer opts.deinit();
    var pipe = try itb.Pipeline.init(
        gpa,
        try util.profileName(arena, env, "streaming-noaead-triple-v1"),
        opts,
    );
    defer pipe.deinit();

    try util.header(init.io);
    const budget = util.minSeconds(env);
    const sizes = [_]usize{ 1 << 20, 16 << 20, 64 << 20 };
    for (sizes) |size| {
        const plain = try gpa.alloc(u8, size);
        defer gpa.free(plain);
        try util.randomFill(plain); // not in the timing loop
        const ctx = Ctx{ .pipe = &pipe, .plain = plain };
        try util.benchCase(init.io, budget, "stream_pump", size, &ctx, runStreamPump);
    }
}
